lib/server.js

```diff
@@ -779,420 +733,306 @@ Server.prototype.toString = function toString() {
+    // Stop
+    if (err === false) {
+        return;
+    }
 
-        // if we have reached this last section of next(), then we are 'done'
-        // with this route.
-        dtrace._rstfy_probes['route-done'].fire(function fire() {
-            return [
-                self.name,
-                route !== null ? route.name : 'pre',
-                id,
-                res.statusCode || 200,
-                res.headers()
-            ];
-        });
+    next();
+};
 
-        // if there is no route, it's because this is running the `pre` handler
-        // chain.
-        if (route === null) {
-            self.emit('preDone', req, res);
-        } else {
-            req.removeListener('close', _requestClose);
-            req.removeListener('aborted', _requestAborted);
-            self.emit('done', req, res, route);
-        }
+/**
+ * After route handlers finished
+ * NOTE: only called when last handler calls next([err])
+ *
+ * @private
+ * @memberof Server
+ * @instance
+ * @function _onHandlerError
+ * @param  {Error|String|undefined} err - router handler error or route name
+ * @param  {Request} req - request
+ * @param  {Response} res - response
+ * @returns {undefined} no return value
+ */
+Server.prototype._onHandlerError = function _onHandlerError(err, req, res) {
+    var self = this;
 
-        // if we have reached here, there are no more handlers in the chain, or
-        // we next(err), and we are done with the request. if errors are still
-        // being emitted (due to being async), skip calling cb now, that will
-        // happen after all error events are done being emitted.
-        if (emittingErrors === false) {
-            return cb(arg);
+    // Call route by name
+    if (_.isString(err)) {
+        var routeName = err;
+        var routeHandler = self.router.lookupByName(routeName, req, res);
+
+        // Cannot find route by name, called when next('route-name') doesn't
+        // find any route, it's a 5xx error as it's a programatic error
+        if (!routeHandler) {
+            var routeByNameErr = new errors.RouteMissingError(
+                "Route by name doesn't exist"
+            );
+            routeByNameErr.code = 'ENOEXIST';
+            self._afterRoute(routeByNameErr, req, res);
+            return;
         }
-
-        // don't really need to return anything, returning null to placate
-        // eslint.
-        return null;
+        routeHandler(req, res, function afterRouter(routeErr) {
+            self._afterRoute(routeErr, req, res);
+        });
+        return;
     }
 
-    var n1 = self._once(next);
-
-    dtrace._rstfy_probes['route-start'].fire(function fire() {
-        return [
-            self.name,
-            route !== null ? route.name : 'pre',
-            id,
-            req.method,
-            req.href(),
-            req.headers
-        ];
-    });
-
-    req.timers = [];
+    // Preserve handler err for finish event
+    res.err = res.err || err;
```

Comment:
I didn't see where this is initially set - what's the use case for storing it on the response directly?

Reply:
It's used when the request succeeds but the response is an error:
https://github.com/restify/node-restify/blob/feat/new-router/lib/server.js#L1093

This is how it's ensured that error is always emitted in `after`.
It's not an elegant solution decorating the `res` object, but in the earlier version it was way to complex, I like that now it's based on the response events.